planner/core/rule_partition_processor.go

```diff
@@ -114,31 +124,50 @@ func (s *partitionProcessor) prune(ds *DataSource) (LogicalPlan, error) {
 		// No result after table pruning.
 		tableDual := LogicalTableDual{RowCount: 0}.Init(ds.context())
 		tableDual.schema = ds.Schema()
-		return tableDual, nil
+		return s.selectOnSomething(sel, tableDual)
 	}
 	if len(children) == 1 {
 		// No need for the union all.
-		return children[0], nil
+		return s.selectOnSomething(sel, children[0])
 	}
 	unionAll := LogicalUnionAll{}.Init(ds.context())
 	unionAll.SetChildren(children...)
 	unionAll.SetSchema(ds.schema)
-	return unionAll, nil
+	return s.selectOnSomething(sel, unionAll)
 }
 
+var solver = expression.NewPartitionPruneSolver()
+
 // canBePrune checks if partition expression will never meets the selection condition.
 // For example, partition by column a > 3, and select condition is a < 3, then canBePrune returns true.
-func (s *partitionProcessor) canBePrune(ctx sessionctx.Context, col *expression.Column, partitionCond expression.Expression, copConds []expression.Expression) (bool, error) {
-	conds := make([]expression.Expression, 0, 1+len(copConds))
+func (s *partitionProcessor) canBePrune(ctx sessionctx.Context, col *expression.Column, partitionCond expression.Expression, rootConds, copConds []expression.Expression) (bool, error) {
+	conds := make([]expression.Expression, 0, 1+len(rootConds)+len(copConds))
 	conds = append(conds, partitionCond)
+	conds = append(conds, rootConds...)
 	conds = append(conds, copConds...)
 	conds = expression.PropagateConstant(ctx, conds)
+	conds = solver.Solve(ctx, conds)
+
+	if len(conds) == 1 {
+		// Constant false.
+		if c, ok := conds[0].(*expression.Constant); ok {
+			if c.GetType().Tp == mysql.TypeTiny {
+				if c.Value.GetInt64() == 0 {
+					return true, nil
+				}
+			}
+		}
+	}
 
 	// Calculate the column range to prune.
-	accessConds := ranger.ExtractAccessConditionsForColumn(conds, col.UniqueID)
-	r, err := ranger.BuildColumnRange(accessConds, ctx.GetSessionVars().StmtCtx, col.RetType)
-	if err != nil {
-		return false, errors.Trace(err)
+	// TODO: Remove prune by calculating range.
+	if col != nil {
+		accessConds := ranger.ExtractAccessConditionsForColumn(conds, col.UniqueID)
```

Comment:
could you add some comment about why we need to calculate the ranges in this place?